### ISSUE TITLE
query: Huge network traffic in thanos multi-cluster setup

### ISSUE THREAD

**Issue by mihailgmihaylov** [improbable-eng/thanos#943]
Thanos image: `improbable/thanos:v0.2.0`
Prometheus image: `quay.io/prometheus/prometheus:v2.3.2`
Go version: `go1.10.5`

I do not know if this may be considered a "bug" or "issue" but mostly I would very much appreciate a clarification about the grpc traffic. Here is our case:

### Background
We have setup a "full-featured" Thanos setup and it has been working great for some time now (thanks for that :) ). It consists of two clusters and all components deployed on them:
* infrastructure cluster:
  - 3x prometheus + thanos sidecar
  - 3x thanos query
  - 1x thanos ruler
  - 1x thanos store
  - 1x thanos compact
* application cluster:
  - 3x prometheus + thanos sidecar
* object storage - one S3 bucket for all (the same objectsore config file for all thanos components).

In the infra cluster, thanos query's options are:
```
        - "query"
        - "--log.level=info"
        - "--cluster.peers=monitoring-thanos-peers:10900"
        - "--store=dns+<thanos_grpc_hostname>:10901"
```
Where `thanos_grpc_hostname` is the dns hostname of a kubernetes LoadBalancer service on the application cluster (the traffic goes through AWS ELB with VPC Peering).

Prometheus settings:
```
          - '--storage.tsdb.retention=12h'
          - '--storage.tsdb.min-block-duration=2h'
          - '--storage.tsdb.max-block-duration=2h'
```

**What happened**
Today we checked the cloudwatch network traffic statistic for the ELB which we use for the thanos query `--store=` option and it turns out that we are transferring through it more than 1T a month.

**What you expected to happen**
As I said, I am not sure if there is a problem. It may be simply that we are just having too much data to transfer. I am afraid that I do not entirely know how Thanos works in that case so I do not know it this is normal or not.

I expected that most of the traffic goes to S3. We have only 12h data on the prometheus nodes so thanos query should not make that huge traffic through the manually configured `--store=`. Thanos store is on the same cluster as Thanos query.

As I was reading the documentation today I came up with three scenarios that thanos-query may produce traffic through the `--store=` (it would be great if you verify):
- when we issue PromQL queries to the Thanos Query API
- when Thanos ruler evaluates alerting rules
- when Thanos ruler creates recording rules

The third one is my biggest suspect right now. We have several recording rules that have a lot of time series and may cause a significant traffic.

**How to reproduce it (as minimally and precisely as possible)**:
I would suggest to simply set a `--store=` option to a load balancer somewhere and track the traffic there. Not sure if it is that easily reproducible though... 

**Full logs to relevant components**

<details>Logs
<p>

The log of one (out of 3) Thanos Query nodes (log level = info):
```
level=info ts=2019-03-19T15:29:51.661577525Z caller=flags.go:90 msg="StoreAPI address that will be propagated through gossip" address=10.*.*.*:10901
level=info ts=2019-03-19T15:29:51.666519423Z caller=flags.go:105 msg="QueryAPI address that will be propagated through gossip" address=10.*.*.*:10902
level=info ts=2019-03-19T15:29:51.679571327Z caller=main.go:256 component=query msg="disabled TLS, key and cert must be set to enable"
level=info ts=2019-03-19T15:29:51.679629136Z caller=query.go:436 msg="starting query node"
level=info ts=2019-03-19T15:29:51.683229998Z caller=query.go:405 msg="Listening for query and metrics" address=0.0.0.0:10902
level=info ts=2019-03-19T15:29:51.683287758Z caller=query.go:428 component=query msg="Listening for StoreAPI gRPC" address=0.0.0.0:10901
level=warn ts=2019-03-19T15:29:51.699229818Z caller=storeset.go:222 component=storeset msg="dropping store, external labels are not unique" address=172.*.*.*:10901
level=warn ts=2019-03-19T15:29:51.69926713Z caller=storeset.go:222 component=storeset msg="dropping store, external labels are not unique" address=172.*.*.*:10901
level=info ts=2019-03-19T15:29:51.699282178Z caller=storeset.go:227 component=storeset msg="adding new store to query storeset" address=172.*.*.*:10901
level=info ts=2019-03-19T15:29:56.704898294Z caller=storeset.go:227 component=storeset msg="adding new store to query storeset" address=10.*.*.*:10901
level=info ts=2019-03-19T15:29:56.704948148Z caller=storeset.go:227 component=storeset msg="adding new store to query storeset" address=10.*.*.*:10901
level=info ts=2019-03-19T15:29:56.704963083Z caller=storeset.go:227 component=storeset msg="adding new store to query storeset" address=10.*.*.*:10901
level=info ts=2019-03-19T15:29:56.704975289Z caller=storeset.go:227 component=storeset msg="adding new store to query storeset" address=172.*.*.*:10901
level=warn ts=2019-03-19T15:29:56.70499892Z caller=storeset.go:222 component=storeset msg="dropping store, external labels are not unique" address=172.*.*.*:10901
level=info ts=2019-03-19T15:29:56.705011534Z caller=storeset.go:227 component=storeset msg="adding new store to query storeset" address=10.*.*.*:10901
level=info ts=2019-03-19T15:29:56.705021858Z caller=storeset.go:227 component=storeset msg="adding new store to query storeset" address=10.*.*.*:10901
level=warn ts=2019-03-19T15:30:01.690481799Z caller=storeset.go:222 component=storeset msg="dropping store, external labels are not unique" address=172.*.*.*:10901
level=info ts=2019-03-19T15:30:06.687400985Z caller=storeset.go:227 component=storeset msg="adding new store to query storeset" address=172.*.*.*:10901
```

</p>
</details>

**Anything else we need to know**

**Environment**:
- OS (e.g. from /etc/os-release):
```
NAME="Ubuntu"
VERSION="18.04.2 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.2 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
- Kernel (e.g. `uname -a`):
```
Linux ip-172-28-0-22 4.15.0-1033-aws #35-Ubuntu SMP Wed Feb 6 13:29:46 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```
- Others:
AWS

**Comment by FUSAKLA** [comment]
Hi, your assumptions are correct AFAIK regarding the traffic.

Just to make sure, query to store does not use ELB? Isn't there possibility that  `dns+<thanos_grpc_hostname>` resolves to ELB address and thus you communicate inside cluster over ELB?  (not familiar witch cloud providers so sorry if I'm completely wrong :) )

Also there is communication with the S3 from the sidecars, store and compactor. Is there any chance any of this uses the ELB also?

**Comment by bwplotka** [comment]
@mihailgmihaylov Thanks for this. 

> I expected that most of the traffic goes to S3. We have only 12h data on the prometheus nodes so thanos query should not make that huge traffic through the manually configured --store=. 

I believe this is something that will solve your problem: https://github.com/improbable-eng/thanos/issues/82

This means that any query that touches now-12h will go through sidecars (your --store things). This is because 12h is still there. Unfortunately there is no simple solution to this, however we can add something that will get us closer to this (: Any suggestions?

> when Thanos ruler evaluates alerting rules
> when Thanos ruler creates recording rules

This might be your thing. In fact we recommend to store your rec or alerting rules locally in Prometheus only if possible. This reduces the traffic majorily (complex rules or alerts done EVERY 20s (eval_interval)) and even reduces risk of network partition.

**Comment by mihailgmihaylov** [comment]
Thank you both for the reply and the suggestions!
Sorry for the delay in answer, but I had to implement some changes and test the result, which took time.

@FUSAKLA My use case is cluster to cluster communication in different VPCs so that's why we chose to use ELB. Hoverer, you are right that we do not need to use load balancer there. I can make the traffic machine to machine traffic (which is cheaper) so this is a good reduction of unnecessary costs.

@bwplotka I analysed the most consuming recording rules and moved them to Prometheus nodes. I judging by the result I think that this was my main consumer of network traffic. I also updated Thanos to the latest version and switched from peers to manually defining thanos store apis (which I forgot to mention that we do not use yet). I think all of these changes helped.

Thank you again for the support.
